File: Labs/Lab8/Solution.py
from typing import List

class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        table = [amount + 1] * (amount + 1)
        table[0] = 0

        for sum in range (1, len(table)):
            for c in coins:
                if c <= sum and table[sum - c] != amount + 1:
                    table[sum] = min(table[sum], table[sum - c] + 1)

        if table[amount] == amount + 1:
            return -1
        else:
            return table[amount]

File: Labs/Lab8/Solution.java
public class Solution {
    public int coinChange(int[] coins, int amount) {
        //minimum number of coins needed to add up to amount
        int[] table = new int[amount + 1];
        for (int i = 0; i < table.length; i++) table[i] = amount + 1;
        table[0] = 0;

        for (int sum = 1; sum < table.length; sum++) {
            for (int c : coins) {
                if (c <= sum && table[sum - c] != amount + 1) {
                    table[sum] = Math.min(table[sum], table[sum - c] + 1);
                }
            }
        }

        return table[amount] == amount + 1 ? -1 : table[amount];
    }
}
